File: MiniDOMOffice/Magazzino/Articolo/CodiciArticolo/CCodiciPerArticolo.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Office

    <Serializable>
    Public Class CCodiciPerArticolo
        Inherits CKeyCollection(Of CodiceArticolo)

        <NonSerialized> Private m_Articolo As Articolo

        Public Sub New()
            Me.m_Articolo = Nothing
        End Sub

        Public Sub New(ByVal articolo As Articolo)
            Me.New()
            Me.Load(articolo)
        End Sub

        Protected Friend Overridable Sub SetArticolo(ByVal articolo As Articolo)
            Me.m_Articolo = articolo
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Articolo IsNot Nothing) Then
                With DirectCast(newValue, CodiceArticolo)
                    .SetArticolo(Me.m_Articolo)
                    .SetOrdine(index)
                End With
            End If
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Articolo IsNot Nothing) Then
                With DirectCast(value, CodiceArticolo)
                    .SetArticolo(Me.m_Articolo)
                    .SetOrdine(index)
                End With
            End If
            MyBase.OnInsert(index, value)
        End Sub

        Protected Friend Overridable Sub Load(ByVal articolo As Articolo)
            If (articolo Is Nothing) Then Throw New ArgumentNullException("articolo")
            Me.Clear()
            Me.m_Articolo = articolo
            If (GetID(articolo) = 0) Then Exit Sub
            Using cursor As New CodiceArticoloCursor()
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                cursor.IDArticolo.Value = GetID(articolo)
                cursor.Ordine.SortOrder = SortEnum.SORT_ASC
                While Not cursor.EOF
                    Me.Add(cursor.Item.Nome, cursor.Item)
                    cursor.MoveNext()
                End While
            End Using
        End Sub

    End Class

End Class

File: MiniDOMOffice/Magazzino/Articolo/CodiciArticolo/Collection/CCodiciPerArticolo.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Store;

namespace minidom
{
    public partial class Store
    {

        /// <summary>
        /// Collezione dei codici articolo
        /// </summary>
        [Serializable]
        public class CCodiciPerArticolo 
            : CKeyCollection<CodiceArticolo>
        {
            [NonSerialized] private Articolo m_Articolo;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CCodiciPerArticolo()
            {
                m_Articolo = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="articolo"></param>
            public CCodiciPerArticolo(Articolo articolo) : this()
            {
                Load(articolo);
            }

            /// <summary>
            /// Imposta l'articolo a cui appartiene la collezione
            /// </summary>
            /// <param name="articolo"></param>
            protected internal virtual void SetArticolo(Articolo articolo)
            {
                m_Articolo = articolo;
                if (articolo is object)
                {
                    foreach (var c in this)
                    {
                        c.SetArticolo(this.m_Articolo);
                    }
                }
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, CodiceArticolo oldValue, CodiceArticolo newValue)
            {
                if (m_Articolo is object)
                {
                    newValue.SetArticolo(m_Articolo);
                    newValue.SetOrdine(index);                     
                }

                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, CodiceArticolo value)
            {
                if (m_Articolo is object)
                {
                    value.SetArticolo(m_Articolo);
                    value.SetOrdine(index);
                }

                base.OnInsert(index, value);
            }

            /// <summary>
            /// Carica da ldb
            /// </summary>
            /// <param name="articolo"></param>
            protected internal virtual void Load(Articolo articolo)
            {
                if (articolo is null)
                    throw new ArgumentNullException("articolo");
                
                Clear();

                m_Articolo = articolo;
                
                if (DBUtils.GetID(articolo, 0) == 0)
                    return;

                using (var cursor = new CodiceArticoloCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.IDArticolo.Value = DBUtils.GetID(articolo);
                    cursor.Ordine.SortOrder = SortEnum.SORT_ASC;
                    while (cursor.Read())
                    {
                        Add(cursor.Item.Nome, cursor.Item);
                    }
                }
            }
        }
    }
}